File: Section10_Probability-for-Robotics/bumperbot_ws/src/bumperbot_examples/src/turtlesim_kinematics.cpp
#include "bumperbot_examples/turtlesim_kinematics.h"


TurtlesimKinematics::TurtlesimKinematics()
{
    turtle1_pose_sub_ = nh_.subscribe("/turtle1/pose", 1000, 
        &TurtlesimKinematics::turtle1PoseCallback, this);
    turtle2_pose_sub_ = nh_.subscribe("/turtle2/pose", 1000, 
        &TurtlesimKinematics::turtle2PoseCallback, this);
}


void TurtlesimKinematics::turtle1PoseCallback(const turtlesim::Pose& pose)
{
    last_turtle1_pose_ = pose;
}


void TurtlesimKinematics::turtle2PoseCallback(const turtlesim::Pose& pose)
{
    last_turtle2_pose_ = pose;
    float Tx = last_turtle2_pose_.x - last_turtle1_pose_.x;
    float Ty = last_turtle2_pose_.y - last_turtle1_pose_.y;
    float theta_rad = last_turtle2_pose_.theta - last_turtle1_pose_.theta;
    float theta_deg = 180 * theta_rad / 3.14;
    ROS_INFO_STREAM("------------------------------------\n" <<
                    "Translation Vector turtle1 -> turtle2 \n" <<
                    "Tx: " << Tx << "\n" <<
                    "Ty: " << Ty << "\n" <<
                    "Rotation Matrix turtle1 -> turtle2 \n" << 
                    "theta (rad): " << theta_rad << "\n" <<
                    "theta (deg): " << theta_deg << "\n" <<
                    "|R11   R12|:  |" << std::cos(theta_rad) << "\t" << -std::sin(theta_rad) << "|\n" <<
                    "|R21   R22|   |" << std::sin(theta_rad) << "\t\t" << std::cos(theta_rad) << "|\n");
}

File: Section10_Probability-for-Robotics/bumperbot_ws/src/bumperbot_examples/src/bumperbot_examples/turtlesim_kinematics.py
#!/usr/bin/env python3
import rospy
from turtlesim.msg import Pose
import math


class TurtlesimKinematics(object):
    
    def __init__(self):
        self.turtle1_pose_sub_ = rospy.Subscriber("/turtle1/pose", Pose, self.turtle1PoseCallback)
        self.turtle2_pose_sub_ = rospy.Subscriber("/turtle2/pose", Pose, self.turtle2PoseCallback)

        self.last_turtle1_pose_ = Pose()
        self.last_turtle2_pose_ = Pose()

    
    def turtle1PoseCallback(self, pose):
        self.last_turtle1_pose_ = pose


    def turtle2PoseCallback(self, pose):
        self.last_turtle2_pose_ = pose
        Tx = self.last_turtle2_pose_.x - self.last_turtle1_pose_.x
        Ty = self.last_turtle2_pose_.y - self.last_turtle1_pose_.y
        theta_rad = self.last_turtle2_pose_.theta - self.last_turtle1_pose_.theta
        theta_deg = 180 * theta_rad / 3.14
        rospy.loginfo("""------------------------------------\n
                      Translation Vector turtle1 -> turtle2\n
                      Tx: %f\n
                      Ty: %f\n
                      Rotation Matrix turtle1 -> turtle2\n 
                      theta (rad): %f\n
                      theta (deg): %f\n
                      |R11   R12|:  |%f %f|\n
                      |R21   R22|   |%f %f|\n""",
                      Tx, Ty, theta_rad, theta_deg,
                      math.cos(theta_rad), -math.sin(theta_rad),
                      math.sin(theta_rad), math.cos(theta_rad)
                      )